invenio_rdm_records/alembic/2186256e8d9b_add_deletion_status_to_rdmrecords.py

```diff
@@ -0,0 +1,76 @@
+#
+# This file is part of Invenio.
+# Copyright (C) 2023 TU Wien.
+#
+# Invenio is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""Add deletion status to RDMRecords."""
+
+import sqlalchemy as sa
+from alembic import op
+
+# revision identifiers, used by Alembic.
+revision = "2186256e8d9b"
+down_revision = "ff860d48fb4b"
+branch_labels = ()
+depends_on = None
+
+
+def _get_default_deletion_status():
+    """Try to get the default value for record deletion status."""
+    try:
+        # first try: get the default value from the DB model class
+        from invenio_rdm_records.records.models import RDMRecordMetadata
+
+        default_value = RDMRecordMetadata.deletion_status.default.arg
+        assert len(default_value) == 1
+
+    except Exception:
+        try:
+            # second try: try to get the value from the enum (more flaky)
+            from invenio_rdm_records.records.systemfields.deletion_status import (
+                RecordDeletionStatusEnum,
+            )
+
+            default_value = RecordDeletionStatusEnum.PUBLISHED.value
+            assert len(default_value) == 1
+
+        except Exception:
+            # fallback: just use 'P', as that's the current default (most flaky)
+            default_value = "P"
+
+    return default_value
+
+
+def upgrade():
+    """Upgrade database."""
+    # step 1: create the columns, but make them nullable for now
+    op.add_column(
+        "rdm_records_metadata",
+        sa.Column("deletion_status", sa.String(length=1), nullable=True),
+    )
+    op.add_column(
+        "rdm_records_metadata_version",
+        sa.Column(
+            "deletion_status",
+            sa.String(length=1),
+            nullable=True,
+        ),
+    )
+
+    # step 2: set default values for existing rows
+    default_value = _get_default_deletion_status()
+    op.execute(f"UPDATE rdm_records_metadata SET deletion_status = '{default_value}'")
+    op.execute(
+        f"UPDATE rdm_records_metadata_version SET deletion_status = '{default_value}'"
+    )
+
+    # step 3: make the original table not nullable
+    op.alter_column("rdm_records_metadata", "deletion_status", nullable=False)
```

Comment:
note: the `rdm_records_metadata_version.deletion_status` table was **not** set `nullable=False` by alembic even before i rewrote the upgrade script.

i'm not sure if it actually matters whether or not we have the column in the `[...]_version` table nullable or not; the critical part is the `rdm_records_metadata` table/column.

Reply:
If you create the database with `invenio db create`, is `rdm_records_metadata_version.deletion_status` set to `nullable=False`? If yes, the alembic recipe should match that.

Reply:
no, with just `invenio db create` it's not marked as `not null`:

```
my-site=# \d rdm_records_metadata_version
                    Table "public.rdm_records_metadata_version"
       Column       |            Type             | Collation | Nullable | Default
--------------------+-----------------------------+-----------+----------+---------
 created            | timestamp without time zone |           |          |
 updated            | timestamp without time zone |           |          |
 id                 | uuid                        |           | not null |
 json               | jsonb                       |           |          |
 version_id         | integer                     |           |          |
 index              | integer                     |           |          |
 bucket_id          | uuid                        |           |          |
 media_bucket_id    | uuid                        |           |          |
 deletion_status    | character varying(1)        |           |          |          <==
 parent_id          | uuid                        |           |          |
 transaction_id     | bigint                      |           | not null |
 end_transaction_id | bigint                      |           |          |
 operation_type     | smallint                    |           | not null |
Indexes:
    "pk_rdm_records_metadata_version" PRIMARY KEY, btree (id, transaction_id)
    "ix_rdm_records_metadata_version_end_transaction_id" btree (end_transaction_id)
    "ix_rdm_records_metadata_version_operation_type" btree (operation_type)
    "ix_rdm_records_metadata_version_transaction_id" btree (transaction_id)

```

i can't find any explicit table definition for it (and think it's automatically created by SQLAlchemy-Continuum?), so i can't tell exactly why it's not marked as `not null` while the "original" table is.

unfortunately i'm not familiar enough with the entire revision/version system to be able to foresee any potential problems here, but i guess it's fine.
some quick manual testing didn't show any issues when editing metadata for existing records and publishing the changes, or creating new versions of records.